File: sdks/node/src/plugins/droid-installation.ts
import path from 'node:path';
import type { InstallConfig } from './base.js';
import {
  AGENT_KEY,
  AUDIT_SCRIPT,
  GUARD_SCRIPT,
  samePath,
} from './droid-contract.js';
import {
  activeDocument,
  installationDocuments,
  sourceDocuments,
  type DroidDocument,
  type DroidSources,
  type RenderedDocument,
} from './droid-config.js';
import { requireHooksEnabled } from './droid-io.js';
import {
  commitManagedInstallation,
  preflightManagedInstallation,
  prepareManagedInstallation,
  type ManagedRuntimePaths,
  type PreparedManagedInstallation,
  type RenameFile,
} from './managed-installation.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
  type PreparedManagedTransaction,
} from './managed-transaction.js';

const DISPLAY_NAME = 'Factory Droid';
const DIRECTORY_LABEL = 'Factory Droid user configuration directory';
const MAX_SOURCE_BYTES = 2 * 1024 * 1024;

export type DroidRuntimePaths = ManagedRuntimePaths;
export type PreparedDroidInstallation = PreparedManagedInstallation;
export type { RenameFile };

function sourceLabel(document: DroidDocument): string {
  if (document.kind === 'settings') return 'Factory Droid user settings';
  if (document.kind === 'local-settings') return 'Factory Droid local settings';
  if (document.kind === 'legacy') return 'Factory Droid legacy hooks';
  return 'Factory Droid user hooks';
}

function hookLocations(sources: DroidSources) {
  return installationDocuments(sources).map((document) => ({
    directoryLabel: document.kind === 'legacy'
      ? 'Factory Droid legacy hooks directory'
      : DIRECTORY_LABEL,
    filePath: document.filePath,
  }));
}

export async function preflightDroidInstallation(
  config: InstallConfig,
  sources: DroidSources,
): Promise<DroidRuntimePaths> {
  requireHooksEnabled(sources);
  return preflightManagedInstallation({
    agentKey: AGENT_KEY,
    hookLocations: hookLocations(sources),
    config,
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
}

function installationChanges(
  sources: DroidSources,
  rendered: readonly RenderedDocument[],
): readonly RenderedDocument[] {
  const target = activeDocument(sources);
  return rendered.filter((item) => item.changed || item.document === target);
}

function sourcePreconditions(
  sources: DroidSources,
  changed: readonly RenderedDocument[],
) {
  const changedPaths = changed.map((item) => item.document.filePath);
  return sourceDocuments(sources)
    .filter((document) => !changedPaths.some((filePath) => samePath(filePath, document.filePath)))
    .map((document) => ({
      filePath: document.filePath,
      label: sourceLabel(document),
      maximumBytes: MAX_SOURCE_BYTES,
      original: document.snapshot,
    }));
}

export async function prepareDroidInstallation(
  config: InstallConfig,
  sources: DroidSources,
  rendered: readonly RenderedDocument[],
): Promise<PreparedDroidInstallation> {
  await preflightDroidInstallation(config, sources);
  const changed = installationChanges(sources, rendered);
  const prepared = await prepareManagedInstallation({
    agentKey: AGENT_KEY,
    displayName: DISPLAY_NAME,
    hookSources: changed.map((item) => ({
      directoryLabel: item.document.kind === 'legacy'
        ? 'Factory Droid legacy hooks directory'
        : DIRECTORY_LABEL,
      label: sourceLabel(item.document),
      filePath: item.document.filePath,
      expectedSource: item.document.exists ? item.document.raw : undefined,
      expectedSnapshot: item.document.snapshot,
      source: item.changed ? item.next : item.document.raw,
    })),
    config,
    auditOptions: { nativePayload: true },
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
  return {
    ...prepared,
    transaction: {
      ...prepared.transaction,
      preconditions: [
        ...sourcePreconditions(sources, changed),
        ...sources.policy.preconditions,
      ],
    },
  };
}

export async function commitDroidInstallation(
  prepared: PreparedDroidInstallation,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedInstallation(prepared, renameFile);
}

export async function prepareDroidUninstall(
  rendered: readonly RenderedDocument[],
): Promise<PreparedManagedTransaction> {
  const changed = rendered.filter((item) => item.changed);
  const changes = await Promise.all(changed.map((item) => prepareManagedFileChange({
    filePath: item.document.filePath,
    label: sourceLabel(item.document),
    next: item.next,
    mode: 0o600,
    expectedSource: item.document.raw,
    expectedSnapshot: item.document.snapshot,
    verifyExpectedSource: true,
  })));
  return {
    displayName: DISPLAY_NAME,
    operation: 'uninstall',
    directories: changed.map((item) => ({
      path: path.dirname(item.document.filePath),
      label: item.document.kind === 'legacy'
        ? 'Factory Droid legacy hooks directory'
        : DIRECTORY_LABEL,
    })),
    changes: changes.filter((change) => change !== undefined),
    preconditions: rendered
      .filter((item) => !item.changed)
      .map((item) => ({
        filePath: item.document.filePath,
        label: sourceLabel(item.document),
        maximumBytes: MAX_SOURCE_BYTES,
        original: item.document.snapshot,
      })),
  };
}

export async function commitDroidUninstall(
  prepared: PreparedManagedTransaction,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedTransaction(prepared, renameFile);
}

File: sdks/go/cmd/elydora/plugins/droid_installation.go
package plugins

import (
	"fmt"
)

type preparedDroidInstallation struct {
	changes       []*fileChange
	preconditions []filePrecondition
	paths         *managedRuntimePaths
}

type preparedDroidUninstall struct {
	changes       []*fileChange
	preconditions []filePrecondition
}

func droidSourceLabel(document *droidDocument) string {
	switch document.kind {
	case "settings":
		return "Factory Droid user settings"
	case "local-settings":
		return "Factory Droid local settings"
	case "legacy":
		return "Factory Droid legacy hooks"
	default:
		return "Factory Droid user hooks"
	}
}

func preflightDroidInstallation(
	config InstallConfig,
	sources *droidSources,
) (*managedRuntimePaths, string, error) {
	if sources == nil || sources.root == nil || sources.policy == nil {
		return nil, "", fmt.Errorf("Factory Droid installation sources are required")
	}
	if err := requireDroidHooksEnabled(sources); err != nil {
		return nil, "", err
	}
	hooks := make([]droidHookSettings, 0, len(droidSourceDocuments(sources)))
	for _, document := range droidSourceDocuments(sources) {
		hooks = append(hooks, document.hooks)
	}
	if err := validateDroidRegexes(hooks...); err != nil {
		return nil, "", err
	}
	if err := validateManagedInstallConfig(config, droidAgentKey, "Factory Droid"); err != nil {
		return nil, "", err
	}
	paths, err := resolveManagedRuntimePaths(config, droidGuardScript, droidAuditScript)
	if err != nil {
		return nil, "", err
	}
	if err := validateManagedRuntimeIdentity(
		paths.agentDirectory, config.AgentID, droidAgentKey, "Factory Droid",
	); err != nil {
		return nil, "", err
	}
	nodePath, err := resolveAbsoluteNodeRuntime("Factory Droid")
	if err != nil {
		return nil, "", err
	}
	return paths, nodePath, nil
}

func prepareRenderedDroidChange(
	rendered *droidRenderedDocument,
) (*fileChange, error) {
	if rendered == nil || !rendered.changed {
		return nil, nil
	}
	return prepareSnapshotSourceChange(
		rendered.document.filePath,
		droidSourceLabel(rendered.document),
		rendered.document.snapshot,
		rendered.next,
		0600,
		rendered.remove,
	)
}

func validateDroidRenderedSet(
	sources *droidSources,
	rendered []*droidRenderedDocument,
) error {
	expected := droidInstallationDocuments(sources)
	if len(rendered) != len(expected) {
		return fmt.Errorf("Factory Droid rendered source set is incomplete")
	}
	for _, document := range expected {
		matches := 0
		for _, item := range rendered {
			if item != nil && item.document != nil &&
				sameManagedPath(item.document.filePath, document.filePath) {
				matches++
			}
		}
		if matches != 1 {
			return fmt.Errorf("Factory Droid rendered source set contains unexpected paths")
		}
	}
	return nil
}

func droidSourcePreconditions(
	sources *droidSources,
	changedPaths []string,
) []filePrecondition {
	preconditions := make([]filePrecondition, 0, len(droidSourceDocuments(sources)))
	for _, document := range droidSourceDocuments(sources) {
		changed := false
		for _, path := range changedPaths {
			if sameManagedPath(document.filePath, path) {
				changed = true
				break
			}
		}
		if !changed {
			preconditions = append(preconditions, filePrecondition{
				filePath:    document.filePath,
				label:       droidSourceLabel(document),
				snapshot:    document.snapshot,
				maximumSize: maxManagedSourceBytes,
			})
		}
	}
	return preconditions
}

func prepareDroidInstallation(
	config InstallConfig,
	sources *droidSources,
	rendered []*droidRenderedDocument,
) (*preparedDroidInstallation, error) {
	paths, _, err := preflightDroidInstallation(config, sources)
	if err != nil {
		return nil, err
	}
	if err := validateDroidRenderedSet(sources, rendered); err != nil {
		return nil, err
	}
	guardScript, auditScript := droidExpectedScripts(config.AgentID, nil)
	changes, err := managedRuntimeFileChanges(
		config, paths, droidAgentKey, string(guardScript), string(auditScript),
	)
	if err != nil {
		return nil, err
	}
	changedPaths := make([]string, 0, len(rendered))
	for _, document := range rendered {
		change, changeErr := prepareRenderedDroidChange(document)
		if changeErr != nil {
			return nil, changeErr
		}
		changes = append(changes, change)
		if change != nil {
			changedPaths = append(changedPaths, change.filePath)
		}
	}
	preconditions := droidSourcePreconditions(sources, changedPaths)
	preconditions = append(preconditions, sources.policy.preconditions...)
	return &preparedDroidInstallation{
		changes:       changes,
		preconditions: preconditions,
		paths:         paths,
	}, nil
}

func writeDroidChanges(
	changes []*fileChange,
	label string,
	rename renameFunc,
	runtimeRoot, agentDirectory string,
	preconditions []filePrecondition,
) error {
	if !hasFileChanges(changes) {
		return writeChanges(changes, label, rename, preconditions...)
	}
	if err := assertFilePreconditions(preconditions, label); err != nil {
		return fmt.Errorf("%s: %w", label, err)
	}
	if agentDirectory != "" {
		if err := EnsurePrivateDirectory(runtimeRoot); err != nil {
			return err
		}
		if err := EnsurePrivateDirectory(agentDirectory); err != nil {
			return err
		}
	}
	return writeChanges(changes, label, rename, preconditions...)
}

func commitDroidInstallation(
	prepared *preparedDroidInstallation,
	rename renameFunc,
) error {
	if prepared == nil || prepared.paths == nil {
		return fmt.Errorf("prepared Factory Droid installation is required")
	}
	return writeDroidChanges(
		prepared.changes,
		"Install Factory Droid hooks",
		rename,
		prepared.paths.runtimeRoot,
		prepared.paths.agentDirectory,
		prepared.preconditions,
	)
}

func prepareDroidUninstall(
	rendered []*droidRenderedDocument,
) (*preparedDroidUninstall, error) {
	changes := make([]*fileChange, 0, len(rendered))
	changedPaths := make([]string, 0, len(rendered))
	for _, item := range rendered {
		change, err := prepareRenderedDroidChange(item)
		if err != nil {
			return nil, err
		}
		changes = append(changes, change)
		if change != nil {
			changedPaths = append(changedPaths, change.filePath)
		}
	}
	preconditions := make([]filePrecondition, 0, len(rendered))
	for _, item := range rendered {
		changed := false
		for _, path := range changedPaths {
			if sameManagedPath(item.document.filePath, path) {
				changed = true
				break
			}
		}
		if !changed {
			preconditions = append(preconditions, filePrecondition{
				filePath:    item.document.filePath,
				label:       droidSourceLabel(item.document),
				snapshot:    item.document.snapshot,
				maximumSize: maxManagedSourceBytes,
			})
		}
	}
	return &preparedDroidUninstall{changes, preconditions}, nil
}

func commitDroidUninstall(
	prepared *preparedDroidUninstall,
	rename renameFunc,
) error {
	if prepared == nil {
		return fmt.Errorf("prepared Factory Droid uninstall is required")
	}
	return writeDroidChanges(
		prepared.changes,
		"Uninstall Factory Droid hooks",
		rename,
		"",
		"",
		prepared.preconditions,
	)
}

func droidExpectedScripts(agentID string, _ map[string]any) ([]byte, []byte) {
	return []byte(generateGuardScript(droidAgentKey, agentID, "", false, "")),
		[]byte(buildHookScriptWithOutput(droidAgentKey, agentID, "", false, true))
}
